### DISCUSSION TITLE
Interface "Implemented Member" Implementation should be smarter

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#1144] [General]
When implementing an interface, I should be able to return members that have a derived type of the specified type.  IE:
````
public interface IInterface {
    IMember Member { get; }
}

public interface IMember {

}

public class Member : IMember {

}

public class MyClass : IInterface {

    //This does not work but should.  Generates CS0738. "Does not implmenent interface member..."
    //public Member Member { get; private set; } = new Member();

    //This does work.
    public IMember Member { get; private set; } = new Member();
}
````

In order to make the above work, I have to do something like this:
````
    public class MyClass : IInterface {
        public Member Member { get; private set; } = new Member();
        IMember IInterface.Member => this.Member;
    }
````

**Comment by DavidArno** [comment]
Why would you want to do this? I can't see any benefit at all to having the class provide a different  implementation signature to the interface.

**Comment by jnm2** [comment]
Duplicate of https://github.com/dotnet/csharplang/issues/49, I believe.

**Comment by TonyValenti** [comment]
Because if you are working with an instance of my interface, I want to provide additional information that the interface itself might not expose.   In order to make the above work, I have to do something like this and it isn't so tidy.  It should just be automatic:
````
    public class MyClass : IInterface {
        public Member Member { get; private set; } = new Member();
        IMember IInterface.Member => this.Member;
    }
````

**Comment by Joe4evr** [comment]
> I have to do something like this and it isn't so tidy.

![](https://i.imgur.com/jks6Qqr.png)
The reason is that if you're accessing your instance *as* the interface, then you have *absolutely* no way to know that additional information exists outside of attempting to cast it.

**Comment by CyrusNajmabadi** [comment]
>  It should just be automatic:

Can you be more precise.  What are the exact rules around which such members are generated?  For example, say i have the following:

```c#
interface IFoo {
     int Foo { get; }
}

class C : IFoo {
    short Foo { get; }
}
```

Should that be allowed?  What about user defined operators?  etc. etc.  In other words, what sorts of conversions should be allowed between the values in the bridge-member and the member the user created?
